pkg/skaffold/debug/util.go

```diff
@@ -0,0 +1,161 @@
+		}
+		return imageConfiguration{}, fmt.Errorf("no build artifact for %q", image)
+	}
+	return applyDebuggingTransforms(l, retriever, registries.DebugHelpersRegistry)
+}
+
+func applyDebuggingTransforms(l manifest.ManifestList, retriever configurationRetriever, debugHelpersRegistry string) (manifest.ManifestList, error) {
+	var updated manifest.ManifestList
+	for _, manifest := range l {
+		obj, _, err := decodeFromYaml(manifest, nil, nil)
+		if err != nil {
+			logrus.Debugf("Unable to interpret manifest for debugging: %v\n", err)
+		} else if transformManifest(obj, retriever, debugHelpersRegistry) {
+			manifest, err = encodeAsYaml(obj)
+			if err != nil {
+				return nil, fmt.Errorf("marshalling yaml: %w", err)
+			}
+			if logrus.IsLevelEnabled(logrus.DebugLevel) {
+				logrus.Debugln("Applied debugging transform:\n", string(manifest))
+			}
+		}
+		updated = append(updated, manifest)
+	}
+
+	return updated, nil
+}
+
+// findArtifact finds the corresponding artifact for the given image.
+// If `builds` is empty, then treat all `image` images as a build artifact.
+func findArtifact(image string, builds []graph.Artifact) *graph.Artifact {
+	if len(builds) == 0 {
+		logrus.Debugf("No build artifacts specified: using image as-is %q", image)
+		return &graph.Artifact{ImageName: image, Tag: image}
+	}
+	for _, artifact := range builds {
+		if image == artifact.ImageName || image == artifact.Tag {
+			logrus.Debugf("Found artifact for image %q", image)
+			return &artifact
+		}
+	}
+	return nil
+}
+
+// retrieveImageConfiguration retrieves the image container configuration for
+// the given build artifact
+func retrieveImageConfiguration(ctx context.Context, artifact *graph.Artifact, insecureRegistries map[string]bool) (imageConfiguration, error) {
+	// TODO: use the proper RunContext
+	apiClient, err := docker.NewAPIClient(&runcontext.RunContext{
+		InsecureRegistries: insecureRegistries,
+	})
+	if err != nil {
+		return imageConfiguration{}, fmt.Errorf("could not connect to local docker daemon: %w", err)
+	}
+
+	// the apiClient will go to the remote registry if local docker daemon is not available
+	manifest, err := apiClient.ConfigFile(ctx, artifact.Tag)
+	if err != nil {
+		logrus.Debugf("Error retrieving image manifest for %v: %v", artifact.Tag, err)
+		return imageConfiguration{}, fmt.Errorf("retrieving image config for %q: %w", artifact.Tag, err)
+	}
+
+	config := manifest.Config
+	logrus.Debugf("Retrieved local image configuration for %v: %v", artifact.Tag, config)
+	// need to duplicate slices as apiClient caches requests
+	return imageConfiguration{
+		artifact:   artifact.ImageName,
+		env:        envAsMap(config.Env),
+		entrypoint: dupArray(config.Entrypoint),
+		arguments:  dupArray(config.Cmd),
+		labels:     dupMap(config.Labels),
+		workingDir: config.WorkingDir,
+	}, nil
+}
+
+// envAsMap turns an array of environment "NAME=value" strings into a map
+func envAsMap(env []string) map[string]string {
```

Comment:
this exists at `github.com/GoogleContainerTools/skaffold/pkg/skaffold/util.EnvMapToSlice`

Reply:
I didn't write this :) just moved it